tailfs/tailfsimpl/compositedav/compositedav.go

```diff
@@ -0,0 +1,228 @@
+	// URL to this.
+	BaseURL string
+
+	// Transport (if specified) is the http transport to use when communicating
+	// with this Child's WebDAV service.
+	Transport http.RoundTripper
+
+	rp       *httputil.ReverseProxy
+	initOnce sync.Once
+}
+
+// CloseIdleConnections forcibly closes any idle connections on this Child's
+// reverse proxy.
+func (c *Child) CloseIdleConnections() {
+	tr, ok := c.Transport.(*http.Transport)
+	if ok {
+		tr.CloseIdleConnections()
+	}
+}
+
+func (c *Child) init() {
+	c.initOnce.Do(func() {
+		c.rp = &httputil.ReverseProxy{
+			Transport: c.Transport,
+			Rewrite:   func(r *httputil.ProxyRequest) {},
+		}
+	})
+}
+
+// Handler implements http.Handler by using a dirfs.FS for showing a virtual
+// read-only folder that represents the Child WebDAV services as sub-folders
+// and proxying all requests for resources on the children to those children
+// via httputil.ReverseProxy instances.
+type Handler struct {
+	// Logf specifies a logging function to use.
+	Logf logger.Logf
+
+	// Clock, if specified, determines the current time. If not specified, we
+	// default to time.Now().
+	Clock tstime.Clock
+
+	// StatCache is an optional cache for PROPFIND results.
+	StatCache *StatCache
+
+	// childrenMu guards children and staticRoot.
+	childrenMu sync.RWMutex
+	children   []*Child
+	staticRoot string
+}
+
+// ServeHTTP implements http.Handler.
+func (h *Handler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if r.Method == "PROPFIND" {
+		h.handlePROPFIND(w, r)
+		return
+	}
+
+	if h.StatCache != nil && r.Method != "GET" {
+		// TODO(oxtoacart): maybe be more selective about invalidating cache
+		// If the user is performing a modification (e.g. PUT, MKDIR, etc),
+		// we need to invalidat the StatCache to make sure we're not knowingly
+		// showing stale stats.
+		h.StatCache.invalidate()
+	}
+
+	mpl := h.maxPathLength(r)
+	pathComponents := shared.CleanAndSplit(r.URL.Path)
+
+	if len(pathComponents) >= mpl {
+		h.delegate(pathComponents[mpl-1:], w, r)
+		return
+	}
+	h.handle(w, r)
+}
+
+// handle handles the request locally using our dirfs.FS.
+func (h *Handler) handle(w http.ResponseWriter, r *http.Request) {
+	h.childrenMu.RLock()
+	children := make([]*dirfs.Child, 0, len(h.children))
+	for _, child := range h.children {
+		children = append(children, &child.Child)
+	}
+	wh := &webdav.Handler{
+		LockSystem: webdav.NewMemLS(),
+		FileSystem: &dirfs.FS{
+			Clock:      h.Clock,
+			Children:   children,
+			StaticRoot: h.staticRoot,
+		},
+	}
+	h.childrenMu.RUnlock()
+
+	wh.ServeHTTP(w, r)
+}
+
+// delegate sends the request to the Child WebDAV server.
+func (h *Handler) delegate(pathComponents []string, w http.ResponseWriter, r *http.Request) string {
+	childName := pathComponents[0]
+	child := h.GetChild(childName)
+	if child == nil {
+		w.WriteHeader(http.StatusNotFound)
+		return childName
+	}
+	u, err := url.Parse(child.BaseURL)
+	if err != nil {
+		w.WriteHeader(http.StatusInternalServerError)
```

Comment:
For cases like this `http.Error(w, err.Error(), http.StatusInternalServerError)` can be useful.

Reply:
Oh that's cool, thanks!